Failsafe is mandatory dependency since RC.6

## Description
Seems we got again the problem with optionality of dependency. 
Build where only auto starter is included fails with NoClassDefFoundError

## Expected Behavior
Default spring starter project with riptide-spring-boot-starter should pass the ContextLoad test
## Actual Behavior
Test fails with 
```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.zalando.riptide.autoconfigure.RiptidePostProcessor]: Factory method 'riptidePostProcessor' threw exception; nested exception is java.lang.NoClassDefFoundError: net/jodah/failsafe/Policy
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:622)
	... 42 more
Caused by: java.lang.NoClassDefFoundError: net/jodah/failsafe/Policy
	at org.zalando.riptide.autoconfigure.RiptideAutoConfiguration.riptidePostProcessor(RiptideAutoConfiguration.java:37)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 43 more
```
## Possible Fix
Check what constructor referenced unconditionally and got Policy as argument.

## Steps to Reproduce
1. Create maven spring boot project via start.spring.io
2. Add riptide-spring-boot-starter as dependency
3. Run ./mvnw test
4.

## Context
<!--- How has this bug affected you? What were you trying to accomplish? -->

## Your Environment
* Version used: RC.6-RC-8
* Link to your project:

--- a/riptide-failsafe/src/main/java/org/zalando/riptide/failsafe/FailsafePlugin.java
+++ b/riptide-failsafe/src/main/java/org/zalando/riptide/failsafe/FailsafePlugin.java
@@ -3,14 +3,12 @@ package org.zalando.riptide.failsafe;
 import lombok.AllArgsConstructor;
 import net.jodah.failsafe.Failsafe;
 import net.jodah.failsafe.Policy;
-import net.jodah.failsafe.RetryPolicy;
 import org.apiguardian.api.API;
 import org.organicdesign.fp.collections.ImList;
 import org.springframework.http.client.ClientHttpResponse;
 import org.zalando.riptide.Plugin;
 import org.zalando.riptide.RequestArguments;
 import org.zalando.riptide.RequestExecution;
-import org.zalando.riptide.idempotency.IdempotencyPredicate;
 
 import java.util.List;
 import java.util.function.Predicate;
@@ -33,22 +31,14 @@ public final class FailsafePlugin implements Plugin {
         this(vec(), identity());
     }
 
-    public FailsafePlugin withPolicy(final BackupRequest<ClientHttpResponse> policy) {
-        return withPolicy(policy, new IdempotencyPredicate());
-    }
-
-    public FailsafePlugin withPolicy(final RetryPolicy<ClientHttpResponse> policy) {
-        return withPolicy(new RetryRequestPolicy(policy));
-    }
-
     public FailsafePlugin withPolicy(final Policy<ClientHttpResponse> policy) {
-        return withPolicy(RequestPolicy.of(policy));
+        return withPolicy(RequestPolicies.of(policy));
     }
 
     public FailsafePlugin withPolicy(
             final Policy<ClientHttpResponse> policy,
             final Predicate<RequestArguments> predicate) {
-        return withPolicy(RequestPolicy.of(policy, predicate));
+        return withPolicy(RequestPolicies.of(policy, predicate));
     }
 
     public FailsafePlugin withPolicy(final RequestPolicy policy) {

--- a/riptide-failsafe/src/main/java/org/zalando/riptide/failsafe/RequestPolicies.java
+++ b/riptide-failsafe/src/main/java/org/zalando/riptide/failsafe/RequestPolicies.java
@@ -1,0 +1,45 @@
+package org.zalando.riptide.failsafe;
+
+import net.jodah.failsafe.Policy;
+import net.jodah.failsafe.RetryPolicy;
+import org.apiguardian.api.API;
+import org.springframework.http.client.ClientHttpResponse;
+import org.zalando.riptide.RequestArguments;
+import org.zalando.riptide.idempotency.IdempotencyPredicate;
+
+import java.util.function.Predicate;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+@API(status = EXPERIMENTAL)
+public final class RequestPolicies {
+
+    private RequestPolicies() {
+
+    }
+
+    public static RequestPolicy of(final Policy<ClientHttpResponse> policy) {
+        if (policy instanceof BackupRequest) {
+            return new ConditionalRequestPolicy(
+                    new DefaultRequestPolicy(policy),
+                    new IdempotencyPredicate());
+        } else if (policy instanceof RetryPolicy) {
+            return new RetryRequestPolicy(
+                    (RetryPolicy<ClientHttpResponse>) policy);
+        }
+
+        return new DefaultRequestPolicy(policy);
+    }
+
+    public static RequestPolicy of(
+            final Policy<ClientHttpResponse> policy,
+            final Predicate<RequestArguments> predicate) {
+        return of(of(policy), predicate);
+    }
+
+    public static RequestPolicy of(
+            final RequestPolicy policy,
+            final Predicate<RequestArguments> predicate) {
+        return new ConditionalRequestPolicy(policy, predicate);
+    }
+}

--- a/riptide-failsafe/src/main/java/org/zalando/riptide/failsafe/RequestPolicy.java
+++ b/riptide-failsafe/src/main/java/org/zalando/riptide/failsafe/RequestPolicy.java
@@ -5,11 +5,9 @@ import org.apiguardian.api.API;
 import org.springframework.http.client.ClientHttpResponse;
 import org.zalando.riptide.RequestArguments;
 
-import java.util.function.Predicate;
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
 
-import static org.apiguardian.api.API.Status.MAINTAINED;
-
-@API(status = MAINTAINED)
+@API(status = EXPERIMENTAL)
 public interface RequestPolicy {
 
     default boolean applies(final RequestArguments arguments) {
@@ -17,21 +15,5 @@ public interface RequestPolicy {
     }
 
     Policy<ClientHttpResponse> prepare(final RequestArguments arguments);
-
-    static RequestPolicy of(final Policy<ClientHttpResponse> policy) {
-        return new DefaultRequestPolicy(policy);
-    }
-
-    static RequestPolicy of(
-            final Policy<ClientHttpResponse> policy,
-            final Predicate<RequestArguments> predicate) {
-        return of(of(policy), predicate);
-    }
-
-    static RequestPolicy of(
-            final RequestPolicy policy,
-            final Predicate<RequestArguments> predicate) {
-        return new ConditionalRequestPolicy(policy, predicate);
-    }
 
 }

--- a/riptide-spring-boot-autoconfigure/pom.xml
+++ b/riptide-spring-boot-autoconfigure/pom.xml
@@ -311,6 +311,7 @@
                                 <include>**/NoRiptideFailsafeTest.java</include>
                             </includes>
                             <classpathDependencyExcludes>org.zalando:riptide-failsafe</classpathDependencyExcludes>
+                            <classpathDependencyExcludes>net.jodah:failsafe</classpathDependencyExcludes>
                         </configuration>
                     </execution>
                 </executions>

--- a/riptide-spring-boot-autoconfigure/src/main/java/org/zalando/riptide/autoconfigure/DefaultRiptideRegistrar.java
+++ b/riptide-spring-boot-autoconfigure/src/main/java/org/zalando/riptide/autoconfigure/DefaultRiptideRegistrar.java
@@ -9,7 +9,6 @@ import io.opentracing.contrib.concurrent.TracedExecutorService;
 import lombok.AllArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import net.jodah.failsafe.CircuitBreaker;
-import net.jodah.failsafe.RetryPolicy;
 import net.jodah.failsafe.Timeout;
 import org.apache.http.client.HttpClient;
 import org.apache.http.client.cache.HttpCacheStorage;
@@ -44,7 +43,6 @@ import org.zalando.riptide.compression.RequestCompressionPlugin;
 import org.zalando.riptide.failsafe.BackupRequest;
 import org.zalando.riptide.failsafe.CircuitBreakerListener;
 import org.zalando.riptide.failsafe.FailsafePlugin;
-import org.zalando.riptide.failsafe.RequestPolicy;
 import org.zalando.riptide.failsafe.TaskDecorator;
 import org.zalando.riptide.httpclient.ApacheClientHttpRequestFactory;
 import org.zalando.riptide.httpclient.metrics.HttpConnectionPoolMetrics;
@@ -403,11 +401,8 @@ final class DefaultRiptideRegistrar implements RiptideRegistrar {
                     () -> {
                         log.debug("Client [{}]: Registering [CircuitBreakerFailsafePlugin]", id);
                         return genericBeanDefinition(FailsafePluginFactory.class)
-                                .setFactoryMethod("create")
-                                .addConstructorArgValue(genericBeanDefinition(RequestPolicy.class)
-                                        .setFactoryMethod("of")
-                                        .addConstructorArgValue(registerCircuitBreaker(id, client))
-                                        .getBeanDefinition())
+                                .setFactoryMethod("createCircuitBreakerPlugin")
+                                .addConstructorArgValue(registerCircuitBreaker(id, client))
                                 .addConstructorArgValue(createTaskDecorator(id, client));
                     });
             return Optional.of(pluginId);
@@ -417,7 +412,7 @@ final class DefaultRiptideRegistrar implements RiptideRegistrar {
 
     private Optional<String> registerRetryPolicyFailsafePlugin(final String id, final Client client) {
         if (client.getRetry().getEnabled()) {
-            final String pluginId = registry.registerIfAbsent(name(id, RetryPolicy.class, FailsafePlugin.class), () -> {
+            final String pluginId = registry.registerIfAbsent(name(id, "RetryPolicy", FailsafePlugin.class), () -> {
                 log.debug("Client [{}]: Registering [RetryPolicyFailsafePlugin]", id);
                 return genericBeanDefinition(FailsafePluginFactory.class)
                         .setFactoryMethod("createRetryFailsafePlugin")
@@ -447,11 +442,8 @@ final class DefaultRiptideRegistrar implements RiptideRegistrar {
                     () -> {
                         log.debug("Client [{}]: Registering [BackupRequestFailsafePlugin]", id);
                         return genericBeanDefinition(FailsafePluginFactory.class)
-                                .setFactoryMethod("create")
-                                .addConstructorArgValue(genericBeanDefinition(FailsafePluginFactory.class)
-                                        .setFactoryMethod("createBackupRequest")
-                                        .addConstructorArgValue(client)
-                                        .getBeanDefinition())
+                                .setFactoryMethod("createBackupRequestPlugin")
+                                .addConstructorArgValue(client)
                                 .addConstructorArgValue(createTaskDecorator(id, client));
                     });
             return Optional.of(pluginId);
@@ -463,12 +455,9 @@ final class DefaultRiptideRegistrar implements RiptideRegistrar {
         if (client.getTimeouts().getEnabled()) {
             final String pluginId = registry.registerIfAbsent(name(id, Timeout.class, FailsafePlugin.class), () -> {
                 log.debug("Client [{}]: Registering [TimeoutFailsafePlugin]", id);
-                final TimeSpan timeout = client.getTimeouts().getGlobal();
                 return genericBeanDefinition(FailsafePluginFactory.class)
-                        .setFactoryMethod("create")
-                        .addConstructorArgValue(RequestPolicy.of(
-                                Timeout.<ClientHttpResponse>of(timeout.toDuration())
-                                        .withCancel(true)))
+                        .setFactoryMethod("createTimeoutPlugin")
+                        .addConstructorArgValue(client)
                         .addConstructorArgValue(createTaskDecorator(id, client));
             });
             return Optional.of(pluginId);

--- a/riptide-spring-boot-autoconfigure/src/main/java/org/zalando/riptide/autoconfigure/FailsafePluginFactory.java
+++ b/riptide-spring-boot-autoconfigure/src/main/java/org/zalando/riptide/autoconfigure/FailsafePluginFactory.java
@@ -2,6 +2,7 @@ package org.zalando.riptide.autoconfigure;
 
 import net.jodah.failsafe.CircuitBreaker;
 import net.jodah.failsafe.RetryPolicy;
+import net.jodah.failsafe.Timeout;
 import net.jodah.failsafe.function.DelayFunction;
 import org.springframework.http.client.ClientHttpResponse;
 import org.zalando.riptide.Plugin;
@@ -13,7 +14,7 @@ import org.zalando.riptide.failsafe.CircuitBreakerListener;
 import org.zalando.riptide.failsafe.CompositeDelayFunction;
 import org.zalando.riptide.failsafe.FailsafePlugin;
 import org.zalando.riptide.failsafe.RateLimitResetDelayFunction;
-import org.zalando.riptide.failsafe.RequestPolicy;
+import org.zalando.riptide.failsafe.RequestPolicies;
 import org.zalando.riptide.failsafe.RetryAfterDelayFunction;
 import org.zalando.riptide.failsafe.RetryException;
 import org.zalando.riptide.failsafe.RetryRequestPolicy;
@@ -21,6 +22,7 @@ import org.zalando.riptide.failsafe.TaskDecorator;
 import org.zalando.riptide.idempotency.IdempotencyPredicate;
 
 import javax.annotation.Nullable;
+import java.time.Duration;
 import java.util.Arrays;
 import java.util.Optional;
 import java.util.concurrent.TimeUnit;
@@ -39,18 +41,45 @@ final class FailsafePluginFactory {
 
     }
 
-    public static Plugin create(
-            final RequestPolicy policy,
+    public static Plugin createCircuitBreakerPlugin(
+            final CircuitBreaker<ClientHttpResponse> breaker,
             final TaskDecorator decorator) {
 
         return new FailsafePlugin()
-                .withPolicy(policy)
+                .withPolicy(breaker)
                 .withDecorator(decorator);
     }
 
-    public static Plugin createRetryFailsafePlugin(
+    public static CircuitBreaker<ClientHttpResponse> createCircuitBreaker(
             final Client client,
-            final TaskDecorator decorator) {
+            final CircuitBreakerListener listener) {
+
+        final CircuitBreaker<ClientHttpResponse> breaker = new CircuitBreaker<>();
+
+        Optional.ofNullable(client.getTimeouts())
+                .filter(RiptideProperties.Timeouts::getEnabled)
+                .map(RiptideProperties.Timeouts::getGlobal)
+                .ifPresent(timeout -> timeout.applyTo(breaker::withTimeout));
+
+        Optional.ofNullable(client.getCircuitBreaker().getFailureThreshold())
+                .ifPresent(threshold -> threshold.applyTo(breaker::withFailureThreshold));
+
+        Optional.ofNullable(client.getCircuitBreaker().getDelay())
+                .ifPresent(delay -> delay.applyTo(breaker::withDelay));
+
+        Optional.ofNullable(client.getCircuitBreaker().getSuccessThreshold())
+                .ifPresent(threshold -> threshold.applyTo(breaker::withSuccessThreshold));
+
+        breaker.withDelay(delayFunction());
+        breaker.onOpen(listener::onOpen);
+        breaker.onHalfOpen(listener::onHalfOpen);
+        breaker.onClose(listener::onClose);
+
+        return breaker;
+    }
+
+    public static Plugin createRetryFailsafePlugin(
+            final Client client, final TaskDecorator decorator) {
 
         final RetryPolicy<ClientHttpResponse> policy = new RetryPolicy<>();
 
@@ -97,48 +126,40 @@ final class FailsafePluginFactory {
                             .withPredicate(new IdempotencyPredicate()))
                     .withPolicy(new RetryRequestPolicy(policy.copy()
                             .handleIf(transientConnectionFaults()))
-                            .withPredicate(alwaysTrue()));
+                            .withPredicate(alwaysTrue()))
+                    .withDecorator(decorator);
 
         } else {
             return new FailsafePlugin()
-                    .withPolicy(new RetryRequestPolicy(policy));
+                    .withPolicy(new RetryRequestPolicy(policy))
+                    .withDecorator(decorator);
         }
     }
 
-    public static CircuitBreaker<ClientHttpResponse> createCircuitBreaker(
-            final Client client,
-            final CircuitBreakerListener listener) {
+    public static Plugin createBackupRequestPlugin(
+            final Client client, final TaskDecorator decorator) {
 
-        final CircuitBreaker<ClientHttpResponse> breaker = new CircuitBreaker<>();
+        final TimeSpan delay = client.getBackupRequest().getDelay();
 
-        Optional.ofNullable(client.getTimeouts())
-                .filter(RiptideProperties.Timeouts::getEnabled)
-                .map(RiptideProperties.Timeouts::getGlobal)
-                .ifPresent(timeout -> timeout.applyTo(breaker::withTimeout));
-
-        Optional.ofNullable(client.getCircuitBreaker().getFailureThreshold())
-                .ifPresent(threshold -> threshold.applyTo(breaker::withFailureThreshold));
-
-        Optional.ofNullable(client.getCircuitBreaker().getDelay())
-                .ifPresent(delay -> delay.applyTo(breaker::withDelay));
-
-        Optional.ofNullable(client.getCircuitBreaker().getSuccessThreshold())
-                .ifPresent(threshold -> threshold.applyTo(breaker::withSuccessThreshold));
-
-        breaker.withDelay(delayFunction());
-        breaker.onOpen(listener::onOpen);
-        breaker.onHalfOpen(listener::onHalfOpen);
-        breaker.onClose(listener::onClose);
-
-        return breaker;
+        return new FailsafePlugin()
+                .withPolicy(RequestPolicies.of(
+                        new BackupRequest<>(
+                                delay.getAmount(),
+                                delay.getUnit()),
+                        new IdempotencyPredicate()))
+                .withDecorator(decorator);
     }
 
-    public static RequestPolicy createBackupRequest(final Client client) {
-        return RequestPolicy.of(
-                new BackupRequest<>(
-                        client.getBackupRequest().getDelay().getAmount(),
-                        client.getBackupRequest().getDelay().getUnit()),
-                new IdempotencyPredicate());
+    public static Plugin createTimeoutPlugin(
+            final Client client, final TaskDecorator decorator) {
+
+        final Duration timeout = client.getTimeouts().getGlobal().toDuration();
+
+        return new FailsafePlugin()
+                .withPolicy(
+                        Timeout.<ClientHttpResponse>of(timeout)
+                                .withCancel(true))
+                .withDecorator(decorator);
     }
 
     private static DelayFunction<ClientHttpResponse, Throwable> delayFunction() {

--- a/riptide-spring-boot-autoconfigure/src/test/java/org/zalando/riptide/autoconfigure/NoRiptideFailsafeTest.java
+++ b/riptide-spring-boot-autoconfigure/src/test/java/org/zalando/riptide/autoconfigure/NoRiptideFailsafeTest.java
@@ -1,22 +1,18 @@
 package org.zalando.riptide.autoconfigure;
 
 import org.junit.jupiter.api.Test;
-import org.springframework.util.ReflectionUtils;
-
-import java.lang.reflect.Method;
 
 import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.springframework.util.ReflectionUtils.getUniqueDeclaredMethods;
 
 final class NoRiptideFailsafeTest {
     @Test
     void shouldReadMicrometerPluginDefinitionWithoutFailsafePlugin() {
-        Method[] methods = ReflectionUtils.getUniqueDeclaredMethods(MicrometerPluginFactory.class);
-        assertNotNull(methods);
+        assertNotNull(getUniqueDeclaredMethods(MicrometerPluginFactory.class));
     }
 
     @Test
     void shouldReadDefaultRiptideRegistrarDefinitionWithoutFailsafePlugin() {
-        Method[] methods = ReflectionUtils.getUniqueDeclaredMethods(DefaultRiptideRegistrar.class);
-        assertNotNull(methods);
+        assertNotNull(getUniqueDeclaredMethods(DefaultRiptideRegistrar.class));
     }
 }

--- a/riptide-spring-boot-autoconfigure/src/test/java/org/zalando/riptide/autoconfigure/NoRiptideFailsafeTest.java
+++ b/riptide-spring-boot-autoconfigure/src/test/java/org/zalando/riptide/autoconfigure/NoRiptideFailsafeTest.java
@@ -9,8 +9,14 @@ import static org.junit.jupiter.api.Assertions.assertNotNull;
 
 final class NoRiptideFailsafeTest {
     @Test
-    void shouldReadClassDefinitionWithoutFailsafePlugin() {
+    void shouldReadMicrometerPluginDefinitionWithoutFailsafePlugin() {
         Method[] methods = ReflectionUtils.getUniqueDeclaredMethods(MicrometerPluginFactory.class);
+        assertNotNull(methods);
+    }
+
+    @Test
+    void shouldReadDefaultRiptideRegistrarDefinitionWithoutFailsafePlugin() {
+        Method[] methods = ReflectionUtils.getUniqueDeclaredMethods(DefaultRiptideRegistrar.class);
         assertNotNull(methods);
     }
 }

